### ISSUE TITLE
Match statement narrowing doesn't work on matchee name shadowing

### ISSUE THREAD

**Issue by MeGaGiGaGon** [astral-sh/ty#4070]
### Summary

If inside a match statement case the code shadows the name being matched on, the match fails to narrow the value
https://play.ty.dev/0420b5f8-eb84-4361-932a-98ee7ec8e64c
```py
from typing import reveal_type
def foo(x: tuple[int, str | None]):
    match x:
        # Expected: no error
        # Actual: Object of type `int` is not assignable to `tuple[int, str | None]` 
        case (x, None):
            # Expected type: `int`
            # Revealed type: `tuple[int, str | None]`
            reveal_type(x)
```

### Version

ty 0.0.62 (df75992e7 2026-07-21) playground 4ce538433

**Comment by carljm** [comment]
Thanks for the report! This is not a bug, it's expected and intentional behavior. The name `x` in the scope of `foo` is already declared with a type of `tuple[int, str | None]`. We allow shadowing a name with a new type declaration, but a match pattern binding is not a type declaration, it's just an assignment. So this code is attempting to assign a value of type `int` to a name already declared as type `tuple[int, str | None]`, just as the diagnostic says.

Once we are recovering from an invalid assignment like that, the subsequent inferred type is really best-effort: we don't know what it should be. In general, in order to avoid cascading errors when the assigned type is wrong, we prefer falling back to the declared type in this case.

Both mypy and pyright reject this code as well.
